mixins/localize-mixin.md

```diff
@@ -0,0 +1,265 @@
+
+## Usage
+
+To use `LocalizeMixin`, apply the mixin then define the namespace base (e.g., `my-component`) in the `constructor()` and implement the following functions (examples in "Language Resources"):
+* `getLanguage(langs)`: returns the page's language based upon `langs`, an array of possible languages determined by `LocalizeMixin` (e.g., `['ar-dz', 'ar', 'en']`)
+* `async getLangResources(lang)`: retrieves the available language resources for `lang`
+
+Consume your web component in a page which has the lang attribute set on the <html> element:
+```html
+<html lang="ar">
+	<head>
+		<script type="module">
+			import '../my-element.js';
+		</script>
+	</head>
+	<body>
+		<my-elem></my-elem>
+	</body>
+</html>
+```
+
+If the language of the page changes (via an update to the `lang` attribute on `<html>`), the mixin will automatically detect the change and re-render the web component. It will fire the event `d2l-localize-behavior-language-changed` when this occurs.
+
+## Language Resources
+
+* Always provide entries for base languages (e.g. "en", "fr", "pt") so that if the user is using a regional language (e.g. "en-gb", "fr-ca", "pt-br") which is missing, it can fall back to the base language.
+* If there's no entry for a particular language, and no base language, the value of `data-lang-default` on the `<html>` element will be used.
+* If no `data-lang-default` is specified, "en" will be used as a last resort.
+
+### Resources in an Object Example
+
+In this example the language resources for `ar` and `en` are stored in the `__langResources` object.
+
+```javascript
+import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize-mixin.js';
+class MyComponent extends LocalizeMixin(LitElement) {
+
+	constructor() {
+		super();
+
+		this.namespaceBase = 'my-component';
+
+		this.__langResources = {
+			'ar': {
+				more: 'المزيد',
+				less: 'أقل'
+			},
+			'en': {
+				more: 'more',
+				less: 'less'
+			}
+		};
+	}
+
+	/*
+	* Gets the current language based upon resources in __langResources
+	* Checks each language in `langs` (e.g., ['ar-dz', 'ar', 'en']) and returns the first language with resources
+	*/
+	getLanguage(langs) {
+		for (let i = 0; i < langs.length; i++) {
+			if (this.__langResources[langs[i]]) {
+				return langs[i];
+			}
+		}
+
+		return null;
+	}
+
+	async getLangResources(lang) {
+		return this.__langResources[lang];
+	}
+
+	render() {
+		return html`<div>${this.localize('more')}</div>`;
+	}
+}
+```
+
+### Dynamically Imported Resources Example
+
+This is more ideal for components with many language resources for each language.
+
+ar.js:
+```javascript
+export const val = {
+	'more': 'المزيد',
+	'less': 'أقل'
+};
+```
+
+component:
+```javascript
+import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize-mixin.js';
+class MyComponent extends LocalizeMixin(LitElement) {
+
+	constructor() {
+		super();
+
+		this.namespaceBase = 'my-component';
```

Comment:
Trying to think of ways it could avoid having to set this entirely. If only the mixin could know the element name (e.g. `d2l-more-less`), but I don't think it has access to that in Lit. In Polymer we had the `is` that was always set. Ah well.

Reply:
Maybe a static getter like how Lit does `properties` and `styles` might be better? I'd call it something like `localizeNamespace` (since `namespaceBase` it kind of generic and many mixins can be used on the same element).